star-rx72n-firmware/lib/rx_fec/src/rx_fec.c

```diff
@@ -0,0 +1,461 @@
+                    dec->survivors[t] &= ~(1ULL << (unsigned)next_state);
+                    if ((state & 1) == 1) {
+                        /* Set the bit if predecessor's LSB is 1 */
+                        dec->survivors[t] |= (1ULL << (unsigned)next_state);
+                    }
+                }
+            }
+        }
+
+        /* Swap path metrics */
+        for (int i = 0; i < k_fec_num_states; i++) {
+            dec->path_metrics[i]     = dec->new_path_metrics[i];
+            dec->new_path_metrics[i] = MAX_PATH_METRIC;
+        }
+    }
+
+    /* Traceback from state 0 (due to tail bits flushing encoder to zero) */
+    /* Use output buffer temporarily for decoded bits (one bit per byte) */
+    /* We'll repack at the end */
+    int state = 0;
+
+    /* Allocate stack-local buffer for decoded bits */
+    /* Maximum symbols = RX_FEC_MAX_SYMBOLS, but we only need num_symbols */
+    /* To avoid VLA, we'll decode in reverse order directly into output */
+
+    size_t data_bits = num_symbols - k_fec_tail_bits;
+    if (data_bits == 0) {
+        return RX_ERR_INVALID_SIZE;
+    }
+
+    size_t output_bytes = (data_bits + 7) / 8;
+
+    /* Clear output buffer */
+    memset(output, 0, output_bytes);
+
+    /* Traceback: work backwards through the trellis */
+    for (size_t t = num_symbols; t > 0; t--) {
+        size_t t_idx = t - 1;
+
+        /* The input bit is the MSB of the current state */
+        uint8_t input_bit = (uint8_t)((state >> (k_fec_constraint_length - 2)) & 1);
+
+        /* Store decoded bit if it's a data bit (not tail bit) */
+        if (t_idx < data_bits) {
+            if (input_bit != 0) {
+                size_t byte_idx = t_idx / 8;
+                size_t bit_pos  = 7 - (t_idx % 8);
+                output[byte_idx] |= (uint8_t)(1 << bit_pos);
+            }
+        }
+
+        /* Get predecessor's LSB from survivors */
+        int predecessor_lsb = (int)((dec->survivors[t_idx] >> state) & 1);
+
+        /* Compute predecessor state: shift left and insert the LSB */
+        state = ((state << 1) & (k_fec_num_states - 1)) | predecessor_lsb;
+    }
+
+    *output_len = output_bytes;
+    return RX_OK;
+}
+
+rx_err_t rx_fec_decode_hard(rx_fec_decoder_t *dec, const uint8_t *data,
+                            size_t data_len, size_t expected_output_len,
+                            uint8_t *output, size_t *output_len)
+{
+    if (dec == NULL || data == NULL || output == NULL || output_len == NULL) {
+        return RX_ERR_INVALID_ARG;
+    }
+
+    if (dec->initialized == 0) {
+        return RX_ERR_INVALID_STATE;
+    }
+
+    if (data_len == 0) {
+        return RX_ERR_INVALID_ARG;
+    }
+
+    /* Convert hard bits to soft bits */
+    size_t num_bits = data_len * 8;
+
+    /* Check if we have enough stack space - use static buffer for safety */
+    /* For production, this should be a caller-provided buffer */
+    static rx_soft_bit_t soft_bits_buffer[RX_FEC_MAX_SYMBOLS * 2];
+
```

Comment:
This variable is static, express why. Also the comment "For production", yeah this code should be setup for prod. This is bad. And verify would this lead to an off by 1 error, should it be `* 2` or `* 2 + 1`.